File: android/features/comparison/src/main/java/com/bolsaterminal/features/comparison/ComparisonViewModel.kt
package com.bolsaterminal.features.comparison

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bolsaterminal.core.common.UiState
import com.bolsaterminal.core.model.OHLCVBar
import com.bolsaterminal.domain.usecase.GetHistoryUseCase
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ComparisonViewModel @Inject constructor(
    private val getHistory: GetHistoryUseCase,
) : ViewModel() {

    companion object {
        val PERIODS = listOf("3mo", "6mo", "1y", "5y", "10y")
    }

    private val _symbolsInput = MutableStateFlow("AAPL, MSFT")
    val symbolsInput: StateFlow<String> = _symbolsInput.asStateFlow()

    private val _period = MutableStateFlow("1y")
    val period: StateFlow<String> = _period.asStateFlow()

    private val _state = MutableStateFlow<UiState<Map<String, List<OHLCVBar>>>>(UiState.Idle)
    val state: StateFlow<UiState<Map<String, List<OHLCVBar>>>> = _state.asStateFlow()
    private var consecutiveRefreshFailures = 0

    fun onSymbolsInputChange(value: String) {
        _symbolsInput.value = value
    }

    fun onPeriodChange(value: String) {
        _period.value = value
        load()
    }

    fun load() {
        val symbols = _symbolsInput.value.split(",").map { it.trim().uppercase() }.filter { it.isNotEmpty() }
        if (symbols.isEmpty()) return

        viewModelScope.launch {
            _state.value = UiState.Loading
            try {
                _state.value = UiState.Loaded(fetch(symbols))
            } catch (e: Exception) {
                _state.value = UiState.Error(e.message.orEmpty())
            }
        }
    }

    /**
     * Periodic silent refresh of the currently selected symbols/period — no
     * loading spinner. After 3 consecutive failures falls back to a full
     * [load] so a sustained outage surfaces as an actionable error instead
     * of staying stuck on stale/empty data forever.
     */
    fun refresh() {
        val symbols = _symbolsInput.value.split(",").map { it.trim().uppercase() }.filter { it.isNotEmpty() }
        if (symbols.isEmpty()) return

        viewModelScope.launch {
            try {
                _state.value = UiState.Loaded(fetch(symbols))
                consecutiveRefreshFailures = 0
            } catch (_: Exception) {
                consecutiveRefreshFailures++
                if (consecutiveRefreshFailures >= 3) {
                    consecutiveRefreshFailures = 0
                    load()
                }
            }
        }
    }

    private suspend fun fetch(symbols: List<String>): Map<String, List<OHLCVBar>> {
        val period = _period.value
        val result = coroutineScope {
            symbols.map { symbol -> async { symbol to getHistory(symbol, period, "1d").bars } }.awaitAll()
        }.toMap()
        // A backend hiccup can return valid-but-empty series rather than
        // actually failing — don't let that silently wipe the comparison.
        val current = (_state.value as? UiState.Loaded<Map<String, List<OHLCVBar>>>)?.data
        val resultHasData = result.values.any { it.isNotEmpty() }
        val currentHasData = current?.values?.any { it.isNotEmpty() } == true
        if (!resultHasData && currentHasData) {
            throw IllegalStateException("Empty comparison response")
        }
        return result
    }
}

File: macos/BolsaTerminal/Features/Comparison/ComparisonViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class ComparisonViewModel {
    enum State {
        case idle, loading, loaded, error(String)
    }

    var symbolsInput: String = "AAPL, MSFT"
    var period: String = "1y"
    var state: State = .idle
    var series: [String: [OHLCVBar]] = [:]
    private var consecutiveRefreshFailures = 0

    func load() async {
        guard !parsedSymbols.isEmpty else { return }
        state = .loading
        do {
            series = try await fetchSeries()
            state = .loaded
        } catch {
            state = .error((error as? LocalizedError)?.errorDescription ?? error.localizedDescription)
        }
    }

    /// Periodic background refresh for the currently selected symbols/period —
    /// no loading flash, silently ignores transient failures. After 3
    /// consecutive failures falls back to a full `load()` so a sustained
    /// outage surfaces as an actionable error instead of staying stuck.
    func refresh() async {
        guard !parsedSymbols.isEmpty else { return }
        do {
            series = try await fetchSeries()
            state = .loaded
            consecutiveRefreshFailures = 0
        } catch {
            consecutiveRefreshFailures += 1
            if consecutiveRefreshFailures >= 3 {
                consecutiveRefreshFailures = 0
                await load()
            }
        }
    }

    private var parsedSymbols: [String] {
        symbolsInput
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespaces).uppercased() }
            .filter { !$0.isEmpty }
    }

    private func fetchSeries() async throws -> [String: [OHLCVBar]] {
        let symbols = parsedSymbols
        let period = self.period
        var result: [String: [OHLCVBar]] = [:]
        try await withThrowingTaskGroup(of: (String, HistoryResponse).self) { group in
            for symbol in symbols {
                group.addTask {
                    let history: HistoryResponse = try await APIClient.shared.request(
                        Endpoints.history(symbol, period: period, interval: "1d")
                    )
                    return (symbol, history)
                }
            }
            for try await (symbol, history) in group {
                result[symbol] = history.bars
            }
        }
        return result
    }
}
